cpp/include/cudf/dictionary/dictionary_column_view.hpp

```diff
@@ -0,0 +1,235 @@
+ * the dictionary_column.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to use for the output column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> set_dictionary( dictionary_column_view const& dictionary_column,
+                                        column_view const& dictionary,
+                                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by merging the specified dictionary elements
+ * from two existing dictionary_columns.
+ *
+ * The indices are updated to appear as though the second dictionary_column is appended to the first.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = {["b","e"],[1,0,1,1,0]}
+ * d3 = merge(d1,d2)
+ * d3 is now {["a","b","c","d","e"],[3,0,2,0,4,1,4,4,1]}
+ * ```
+ *
+ * @throw cudf_logic_error if the dictionary types do not match for both dictionary columns.
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param dictionary New keys to remove from the dictionary_column
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> merge_dictionaries( dictionary_column_view const& dictionary_column1,
+                                            dictionary_column_view const& dictionary_column2,
+                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a new dictionary column by copying the dictionary from the provided
+ * dictionary_column and setting the indices from the provided indices parameter.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * d2 = gather(d1,[1,0,0,2,2])
+ * d2 is now {["a","c","d"],[1,0,0,2,2]}
+ * ```
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param indices Indices to use for the new dictionary column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New dictionary column.
+ */
+std::unique_ptr<column> gather( dictionary_column_view const& dictionary_column,
+                                column_view const& indices,
+                                rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+/**
+ * @brief Create a column by gathering the elements from the provided
+ * dictionary_column using the specified indices.
+ *
+ * ```
+ * d1 = {["a","c","d"],[2,0,1,0]}
+ * s = gather_type(d1,[1,0,0,2,2])
+ * s is now ["c","a","a","d","d"]
+ * ```
+ *
+ * @param dictionary_column Existing dictionary column.
+ * @param indices Indices to use for the new dictionary column.
+ * @param mr Resource for allocating memory for the output.
+ * @return New column with type matching the dictionary_column's dictionary.
+ */
+std::unique_ptr<column> gather_type( dictionary_column_view const& dictionary_column,
```

Comment:
Isn't this the same as what would happen if we passed `dictionary_column` as the input to `cudf::gather`? Or would that return a dictionary?

We should try to have one way to gather in cudf.